File: SonarApp/app/src/main/java/com/golovkobalak/sonarapp/service/LocationHelper.java
package com.golovkobalak.sonarapp.service;

import android.content.Context;
import android.location.Location;
import android.os.Looper;

import android.util.Log;
import com.google.android.gms.location.FusedLocationProviderClient;
import com.google.android.gms.location.LocationCallback;
import com.google.android.gms.location.LocationRequest;
import com.google.android.gms.location.LocationResult;
import com.google.android.gms.location.LocationServices;

public class LocationHelper {

    public static Location CURR_LOCATION; // Static field to store the latest location
    private FusedLocationProviderClient fusedLocationClient;
    private LocationRequest locationRequest;
    private LocationCallback locationCallback;

    public LocationHelper(Context context) {
        fusedLocationClient = LocationServices.getFusedLocationProviderClient(context);
        createLocationRequest();
        createLocationCallback();
    }

    private void createLocationRequest() {
        locationRequest = LocationRequest.create();
        locationRequest.setInterval(1000); // Update interval in milliseconds (e.g., every 1 seconds)
        locationRequest.setFastestInterval(500); // Fastest update interval in milliseconds
        locationRequest.setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY); // High accuracy mode
    }

    private void createLocationCallback() {
        locationCallback = new LocationCallback() {
            @Override
            public void onLocationResult(LocationResult locationResult) {
                if (locationResult != null) {
                    Location location = locationResult.getLastLocation();
                    if (location != null) {
                        CURR_LOCATION = location; // Update the static field with the latest location
                    }
                }
            }
        };
    }

    public void start() {
        fusedLocationClient.requestLocationUpdates(locationRequest, locationCallback, Looper.getMainLooper());
    }

    public void stopLocationUpdates() {
        fusedLocationClient.removeLocationUpdates(locationCallback);
    }

    // Method to get the last known location from the static field
    public static Location getLastLocation() {
        return CURR_LOCATION;
    }
}

File: Android/app/src/main/java/com/gmail/golovkobalak/sonar/service/LocationHelper.kt
package com.gmail.golovkobalak.sonar.service

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.location.Location
import android.os.Looper
import android.util.Log
import androidx.core.app.ActivityCompat
import com.google.android.gms.location.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlin.math.*

object LocationHelper {
    private var fusedLocationClient: FusedLocationProviderClient? = null
    private var locationRequest: LocationRequest? = null
    private var locationCallback: LocationCallback? = null
    private val _lastLocation = MutableStateFlow(Location("KyivCenter").apply {
        latitude = 50.4501 // Kyiv's latitude
        longitude = 30.5234 // Kyiv's longitude
        altitude = 0.0 // Altitude in meters
        bearing = 0.0f // Bearing in degrees
        speed = 0.0f // Speed in meters per second
    })
    val lastLocation: StateFlow<Location> = _lastLocation

    // Update the location when it changes
    fun updateLocation(location: Location) {
        Log.d(this.javaClass.name, "location is updated: " + location)
        _lastLocation.value = location
    }

    init {
        createLocationRequest()
        createLocationCallback()
    }

    private fun createLocationRequest() {
        locationRequest = LocationRequest.Builder(100)
            .setMaxUpdateAgeMillis(500)
            .setPriority(Priority.PRIORITY_HIGH_ACCURACY)
            .build()
    }

    private fun createLocationCallback() {
        locationCallback = object : LocationCallback() {
            override fun onLocationResult(locationResult: LocationResult) {
                val location = locationResult.lastLocation
                if (location != null && hasLocationChangedByMeters(lastLocation.value, location, 2.0)) {
                    updateLocation(location)
                }
            }
        }
    }

    private fun haversineDistance(location1: Location, location2: Location): Double {
        val lat1 = Math.toRadians(location1.latitude)
        val lon1 = Math.toRadians(location1.longitude)
        val lat2 = Math.toRadians(location2.latitude)
        val lon2 = Math.toRadians(location2.longitude)

        val dLat = lat2 - lat1
        val dLon = lon2 - lon1

        val a = sin(dLat / 2).pow(2) + cos(lat1) * cos(lat2) * sin(dLon / 2).pow(2)
        val c = 2 * asin(sqrt(a))

        // Radius of the Earth in meters (mean value)
        val earthRadius = 6371000.0

        return earthRadius * c
    }

    fun hasLocationChangedByMeters(location1: Location, location2: Location, meters: Double): Boolean {
        val distance = haversineDistance(location1, location2)
        return distance >= meters
    }

    fun start(context: Context): Boolean {
        if (ActivityCompat.checkSelfPermission(
                context,
                Manifest.permission.ACCESS_FINE_LOCATION
            ) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(
                context, Manifest.permission.ACCESS_COARSE_LOCATION
            ) != PackageManager.PERMISSION_GRANTED
        ) {
            Log.d(javaClass.name, "No permission to start")
            return false
        }
        fusedLocationClient = LocationServices.getFusedLocationProviderClient(context)
        fusedLocationClient!!.requestLocationUpdates(locationRequest!!, locationCallback!!, Looper.getMainLooper())
        return true
    }

    fun stopLocationUpdates() {
        fusedLocationClient!!.removeLocationUpdates(locationCallback!!)
    }
}